### PR TITLE
mfs redo

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is an extraction of the refactoring that i did to the core 'ipnsfs' code, applied on top of dev0.4.0

### CODE REVIEW

File: fuse/ipns/ipns_unix.go [hunk 1]
```diff
@@ -121,12 +169,8 @@ func (s *Root) Lookup(ctx context.Context, name string) (fs.Node, error) {
 		return nil, fuse.ENOENT
 	}
 
-	// Local symlink to the node ID keyspace
-	if name == "local" {
```
Comment (jbenet) [inline fuse/ipns/ipns_unix.go]: yay not having to special case

File: mfs/dir.go [hunk 2]
```diff
@@ -168,15 +170,51 @@ func (d *Directory) childUnsync(name string) (FSNode, error) {
...
+		}
+
+		k, err := nd.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		child.Hash = k.B58String()
+
+		out = append(out, child)
```
Comment (jbenet) [inline mfs/dir.go]: maybe once IPLD lands we can make unixfs dirs store the sizes too to avoid having to read the entire next layer in order to `ls`....

File: mfs/system.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	repub *Republisher
+
+	dserv dag.DAGService
+
+	Type string
+}
+
+type PubFunc func(context.Context, key.Key) error
+
+// newRoot creates a new Root for the given key, and starts up a republisher routine
```
Comment (jbenet) [inline mfs/system.go]: comment outdated

File: mfs/system.go [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+	Lock()
+	Unlock()
+}
+
+// Root represents the root of a filesystem tree pointed to by a given keypair
+type Root struct {
+	// node is the merkledag node pointed to by this keypair
+	node *dag.Node
+
+	// val represents the node pointed to by this key. It can either be a File or a Directory
```
Comment (jbenet) [inline mfs/system.go]: comment outdated

File: mfs/system.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+type FSNode interface {
+	GetNode() (*dag.Node, error)
+	Type() NodeType
+	Lock()
+	Unlock()
+}
+
+// Root represents the root of a filesystem tree pointed to by a given keypair
+type Root struct {
+	// node is the merkledag node pointed to by this keypair
```
Comment (jbenet) [inline mfs/system.go]: outdated

File: mfs/system.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+// FSNode represents any node (directory, root, or file) in the ipns filesystem
+type FSNode interface {
+	GetNode() (*dag.Node, error)
+	Type() NodeType
+	Lock()
+	Unlock()
+}
+
+// Root represents the root of a filesystem tree pointed to by a given keypair
```
Comment (jbenet) [inline mfs/system.go]: here too! wheee

File: fuse/ipns/ipns_unix.go (hunk 1 above)
Comment (whyrusleeping) [inline fuse/ipns/ipns_unix.go]: yeap, that was bugging me for a while.

File: mfs/system.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+type NodeType int
+
+const (
+	TFile NodeType = iota
+	TDir
+)
+
+// FSNode represents any node (directory, root, or file) in the ipns filesystem
```
Comment (jbenet) [inline mfs/system.go]: s/ipns/mfs

File: mfs/dir.go (hunk 2 above)
Comment (whyrusleeping) [inline mfs/dir.go]: :+1:

File: mfs/system.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+// for it
+func NewRoot(parent context.Context, ds dag.DAGService, node *dag.Node, pf PubFunc) (*Root, error) {
+	ndk, err := node.Key()
+	if err != nil {
+		return nil, err
+	}
+
+	root := &Root{
+		node:  node,
+		repub: NewRepublisher(parent, pf, time.Millisecond*300, time.Second*3),
```
Comment (jbenet) [inline mfs/system.go]: wonder if this -- the published coalescing -- should happen in ipns, with mfs _always_ calling the function

Comment (jbenet) [inline mfs/system.go]: could be a different PR, too.

File: unixfs/mod/dagmodifier_test.go [hunk 9]
```diff
@@ -415,164 +411,92 @@ func TestDagTruncate(t *testing.T) {
...
 		t.Fatal(err)
 	}
-	for range out {
+
+	if size != 10 {
+		t.Fatal("size was incorrect!")
 	}
 }
 
-func TestCorrectPinning(t *testing.T) {
```
Comment (jbenet) [inline unixfs/mod/dagmodifier_test.go]: pinning test totally gone?

**Comment by jbenet** [comment]
few comments, but LGTM

File: unixfs/mod/dagmodifier_test.go (hunk 9 above)
Comment (whyrusleeping) [inline unixfs/mod/dagmodifier_test.go]: since we dont do indirect pinning, the dagmodifier doesnt need to be responsible for any sort of pinning anymore, its great!

File: mfs/system.go (hunk 8 above)
Comment (whyrusleeping) [inline mfs/system.go]: i'm not opposed to that. But i also couldnt think of any applications that really cared about every single intermediate state. (vim wrote a .swp file, vim changed the swap file because you pressed a key, vim changed the swap file because... ad nauseum)

File: unixfs/mod/dagmodifier_test.go (hunk 9 above)
Comment (jbenet) [inline unixfs/mod/dagmodifier_test.go]: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers: :dancers:

File: mfs/system.go (hunk 8 above)
Comment (jbenet) [inline mfs/system.go]: leave it until we need to

**Comment by whyrusleeping** [comment]
@jbenet comments fixed
